<a href="https://colab.research.google.com/github/robertdc2/IS-170-Machine-Learning/blob/5.-Association-Rules/Sanchez_Valenzuela_Saligan_Lab5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Association Rules**

# Using 'GroceryStoreDataSet.CSV'

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules


In [ ]:
from google.colab import files
uploaded = files.upload()

Saving GroceryStoreDataSet.csv to GroceryStoreDataSet (1).csv


In [ ]:
#Since it's missing a header we'll create one, the new column header 'Products' 
df=pd.read_csv('GroceryStoreDataSet.csv', names=['products'])
df.head() 


products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [ ]:
df.shape

(20, 1)

In [ ]:
#the following code converts a pandas dataframe into a list of lists.
data=list(df["products"].apply(lambda x:x.split(",")))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [ ]:
#Let's transform the list with one-hot encoding (make sure you understand the one-hot encoding concept first)
#Each column represents an item and each row represent a record or a transaction. 
#Initially, if that particular item occurs then it's labeled TRUE, if not then FALSE.
#After that, TRUE is converted to 1 and FALSE to 0 because computer loves numeric values.

from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 & 1 for easy readability
df=df.replace(False,0)
df=df.replace(True,1)

df
                

BISCUIT  BOURNVITA  BREAD  COCK  COFFEE  CORNFLAKES  JAM  MAGGI  MILK  \
0         1          0      1     0       0           0    0      0     1   
1         1          0      1     0       0           1    0      0     1   
2         0          1      1     0       0           0    0      0     0   
3         0          0      1     0       0           0    1      1     1   
4         1          0      0     0       0           0    0      1     0   
5         0          1      1     0       0           0    0      0     0   
6         0          0      0     0       0           1    0      1     0   
7         1          0      1     0       0           0    0      1     0   
8         0          0      1     0       0           0    1      1     0   
9         0          0      1     0       0           0    0      0     1   
10        1          0      0     1       1           1    0      0     0   
11        1          0      0     1       1           1    0      0     0   
12        0          1      0     0       1           0    0      0     0   
13        0          0      1     1       1           0    0      0     0   
14        1          0      1     0       0           0    0      0     0   
15        0          0      0     0       1           1    0      0     0   
16        0          1      1     0       0           0    0      0     0   
17        0          0      1     0       1           0    0      0     0   
18        0          0      1     0       1           0    0      0     0   
19        0          0      0     0       1           1    0      0     1   

    SUGER  TEA  
0       0    0  
1       0    0  
2       0    1  
3       0    0  
4       0    1  
5       0    1  
6       0    1  
7       0    1  
8       0    1  
9       0    0  
10      0    0  
11      0    0  
12      1    0  
13      0    0  
14      1    0  
15      1    0  
16      1    0  
17      1    0  
18      1    0  
19      0    1

In [ ]:
#We'll try 0.2 for the SUPPORT threshold value (but you can set it to any value you want).
# The parameter (use_colnames=True) is used to convert the integer back to its respective name.

df=apriori(df,min_support=0.2,use_colnames=True)
df

support              itemsets
0      0.35             (BISCUIT)
1      0.20           (BOURNVITA)
2      0.65               (BREAD)
3      0.40              (COFFEE)
4      0.30          (CORNFLAKES)
5      0.25               (MAGGI)
6      0.25                (MILK)
7      0.30               (SUGER)
8      0.35                 (TEA)
9      0.20      (BISCUIT, BREAD)
10     0.20         (BREAD, MILK)
11     0.20        (SUGER, BREAD)
12     0.20          (BREAD, TEA)
13     0.20  (COFFEE, CORNFLAKES)
14     0.20       (COFFEE, SUGER)
15     0.20          (MAGGI, TEA)

With the SUPPORT values above, you can try and match your own manual calculation. You notice how some are grouped with two. ALL items have been tested alone and in groups, but only these ones with the SUPPORT threshold 0.2 or higher values (i.e. other groups didn't make 0.2 therefore not shown).

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0       (BREAD)        (MILK)                0.65                0.25   
1        (MILK)       (BREAD)                0.25                0.65   
2       (SUGER)       (BREAD)                0.30                0.65   
3       (BREAD)       (SUGER)                0.65                0.30   
4      (COFFEE)  (CORNFLAKES)                0.40                0.30   
5  (CORNFLAKES)      (COFFEE)                0.30                0.40   
6      (COFFEE)       (SUGER)                0.40                0.30   
7       (SUGER)      (COFFEE)                0.30                0.40   
8       (MAGGI)         (TEA)                0.25                0.35   
9         (TEA)       (MAGGI)                0.35                0.25   

   support  confidence      lift  leverage  conviction  
0      0.2    0.307692  1.230769    0.0375    1.083333  
1      0.2    0.800000  1.230769    0.0375    1.750000  
2      0.2    0.666667  1.025641    0.0050    1.050000  
3      0.2    0.307692  1.025641    0.0050    1.011111  
4      0.2    0.500000  1.666667    0.0800    1.400000  
5      0.2    0.666667  1.666667    0.0800    1.800000  
6      0.2    0.500000  1.666667    0.0800    1.400000  
7      0.2    0.666667  1.666667    0.0800    1.800000  
8      0.2    0.800000  2.285714    0.1125    3.250000  
9      0.2    0.571429  2.285714    0.1125    1.750000

From above result, we say that the support is 0.2 calculated by dividing the number of transactions containing milk and bread by the total number of transactions.  The confidence level is 0.30 shows that out of all the transactions that contain bread 0.30 contain milk too.  The lift 1.23 tells us that milk is 1.23 times more likely to be bought by the customer who also buys bread.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents consequents  antecedent support  consequent support  support  \
0        (MILK)     (BREAD)                0.25                0.65      0.2   
1       (SUGER)     (BREAD)                0.30                0.65      0.2   
2  (CORNFLAKES)    (COFFEE)                0.30                0.40      0.2   
3       (SUGER)    (COFFEE)                0.30                0.40      0.2   
4       (MAGGI)       (TEA)                0.25                0.35      0.2   

   confidence      lift  leverage  conviction  
0    0.800000  1.230769    0.0375        1.75  
1    0.666667  1.025641    0.0050        1.05  
2    0.666667  1.666667    0.0800        1.80  
3    0.666667  1.666667    0.0800        1.80  
4    0.800000  2.285714    0.1125        3.25

From above result, let's take a look at the first row items: milk and bread. 
The two items have the support of 0.20, where milk is 0.25 and bread is 0.65. 
confidence is 0.8, this implies that if a person bought milk then there is 80% bread is also bought. There are also positive correlations as we can see from the positive values of lift, leverage, and conviction.